File: Sources/MutsumiPet/Stores/PetStore.swift
import Foundation

@MainActor
final class PetStore: ObservableObject {
    @Published private(set) var mood: PetMood = .sleepy
    @Published private(set) var pose: PetPose = .idle
    @Published private(set) var message = PetDialogueScene.idle.lines[0]
    @Published var showsBubble = true
    @Published private(set) var layerMode: WindowLayerMode
    @Published private(set) var scale: Double
    @Published private(set) var activity: PetActivity = .idle
    @Published private(set) var walkingDirection = -1
    @Published private(set) var wanderingEnabled: Bool
    @Published private(set) var wanderSpeed: Double
    @Published private(set) var animationFrame = 0

    private var dismissTask: Task<Void, Never>?
    private var settleTask: Task<Void, Never>?
    private let defaults: UserDefaults
    private var activityTask: Task<Void, Never>?
    private var animationSequenceIndex = 0
    private var walkingDistance: Double = 0

    private static let scaleKey = "pet.scale"
    private static let layerModeKey = "pet.windowLayerMode"
    private static let wanderingEnabledKey = "pet.wanderingEnabled"
    private static let wanderSpeedKey = "pet.wanderSpeed"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        let savedScale = defaults.object(forKey: Self.scaleKey) as? Double ?? 1
        scale = min(max(savedScale, 0.6), 1.4)
        layerMode = WindowLayerMode(
            rawValue: defaults.string(forKey: Self.layerModeKey) ?? ""
        ) ?? .front
        wanderingEnabled = defaults.object(forKey: Self.wanderingEnabledKey) as? Bool ?? true
        let savedSpeed = defaults.object(forKey: Self.wanderSpeedKey) as? Double ?? 42
        wanderSpeed = min(max(savedSpeed, 18), 90)
    }

    func reactToTap(randomDialogueIndex: Int? = nil) {
        interruptActivity()
        mood = PetMood.allCases[(mood.rawValue + 1) % PetMood.allCases.count]
        pose = switch mood {
        case .sleepy: .idle
        case .curious: .curious
        case .pleased: .happy
        }
        message = dialogueScene(for: pose).line(at: randomDialogueIndex)
        showsBubble = true
        scheduleDismiss()
    }

    func idleTick(randomIndex: Int? = nil, sleeping: Bool? = nil) {
        guard showsBubble == false else { return }
        let usesSleepingPose = sleeping ?? Bool.random()
        let scene: PetDialogueScene = usesSleepingPose ? .sleeping : .idle
        message = scene.line(at: randomIndex)
        mood = .sleepy
        pose = usesSleepingPose ? .sleeping : .idle
        showsBubble = true
        scheduleDismiss()
    }

    func beginDrag(randomDialogueIndex: Int? = nil) {
        guard pose != .grabbed else { return }
        interruptActivity()
        settleTask?.cancel()
        dismissTask?.cancel()
        pose = .grabbed
        mood = .curious
        message = PetDialogueScene.grabbed.line(at: randomDialogueIndex)
        showsBubble = true
    }

    func endDrag(randomDialogueIndex: Int? = nil) {
        guard pose == .grabbed else { return }
        pose = .curious
        message = PetDialogueScene.released.line(at: randomDialogueIndex)
        scheduleDismiss()
        settleTask?.cancel()
        settleTask = Task { @MainActor [weak self] in
            try? await Task.sleep(for: .milliseconds(120))
            guard Task.isCancelled == false, self?.pose != .grabbed else { return }
            self?.pose = .idle
            self?.mood = .sleepy
        }
    }

    func toggleBubble() {
        showsBubble.toggle()
        if showsBubble { scheduleDismiss() }
    }

    func setScale(_ newScale: Double) {
        scale = min(max(newScale, 0.6), 1.4)
        defaults.set(scale, forKey: Self.scaleKey)
    }

    func setLayerMode(_ mode: WindowLayerMode) {
        layerMode = mode
        defaults.set(mode.rawValue, forKey: Self.layerModeKey)
    }

    func setWanderingEnabled(_ enabled: Bool) {
        wanderingEnabled = enabled
        defaults.set(enabled, forKey: Self.wanderingEnabledKey)
        if enabled == false { interruptActivity() }
    }

    func setWanderSpeed(_ speed: Double) {
        wanderSpeed = min(max(speed, 18), 90)
        defaults.set(wanderSpeed, forKey: Self.wanderSpeedKey)
    }

    func lifestyleTick(randomEvent: Int? = nil, randomDialogueIndex: Int? = nil) {
        guard wanderingEnabled, pose != .grabbed, activity == .idle else { return }
        switch randomEvent ?? Int.random(in: 0..<8) {
        case 0, 1, 2, 3:
            performLifestyle(.walking, randomDialogueIndex: randomDialogueIndex)
        case 4:
            performLifestyle(.drinkingTea, randomDialogueIndex: randomDialogueIndex)
        case 5:
            performLifestyle(.eatingSnack, randomDialogueIndex: randomDialogueIndex)
        case 6:
            pose = .sleeping
            mood = .sleepy
            message = PetDialogueScene.sleeping.line(at: randomDialogueIndex)
            showsBubble = true
            scheduleDismiss()
        default:
            pose = .curious
            mood = .curious
            message = PetDialogueScene.curious.line(at: randomDialogueIndex)
            showsBubble = true
            scheduleDismiss()
        }
    }

    func performLifestyle(_ requestedActivity: PetActivity, randomDialogueIndex: Int? = nil) {
        guard wanderingEnabled, pose != .grabbed else { return }
        interruptActivity()
        switch requestedActivity {
        case .walking:
            // Arrival normally ends walking; this timeout is only a safety net
            // for a window that cannot move or disappears mid-activity.
            beginActivity(.walking, duration: .seconds(20))
            pose = .curious
            mood = .curious
            message = PetDialogueScene.walking.line(at: randomDialogueIndex)
            showsBubble = true
            scheduleDismiss()
        case .drinkingTea:
            beginActivity(.drinkingTea, duration: .seconds(7))
            pose = .idle
            mood = .pleased
            message = PetDialogueScene.drinkingTea.line(at: randomDialogueIndex)
            showsBubble = true
            scheduleDismiss()
        case .eatingSnack:
            beginActivity(.eatingSnack, duration: .seconds(7))
            pose = .happy
            mood = .pleased
            message = PetDialogueScene.eatingSnack.line(at: randomDialogueIndex)
            showsBubble = true
            scheduleDismiss()
        case .idle:
            break
        }
    }

    func updateWalkingDirection(_ direction: Int) {
        walkingDirection = direction < 0 ? -1 : 1
    }

    func advanceActivityFrame() {
        guard activity != .idle else { return }
        let sequence = activity.frameSequence
        animationSequenceIndex = (animationSequenceIndex + 1) % sequence.count
        animationFrame = sequence[animationSequenceIndex]
    }

    func advanceWalkingFrame(distance: Double) {
        guard activity == .walking, distance > 0 else { return }
        walkingDistance += distance
        let strideLength = 44 * scale
        let phase = walkingDistance.truncatingRemainder(dividingBy: strideLength) / strideLength
        animationFrame = min(Int(phase * 4), 3)
    }

    func finishWalking() {
        guard activity == .walking else { return }
        interruptActivity()
        pose = .idle
        mood = .sleepy
        message = PetDialogueScene.idle.line()
        showsBubble = true
        scheduleDismiss()
    }

    func speakForCurrentState(randomDialogueIndex: Int? = nil) {
        message = currentDialogueScene.line(at: randomDialogueIndex)
        showsBubble = true
        scheduleDismiss()
    }

    private var currentDialogueScene: PetDialogueScene {
        switch activity {
        case .walking: .walking
        case .drinkingTea: .drinkingTea
        case .eatingSnack: .eatingSnack
        case .idle: dialogueScene(for: pose)
        }
    }

    private func dialogueScene(for pose: PetPose) -> PetDialogueScene {
        switch pose {
        case .idle: .idle
        case .curious: .curious
        case .happy: .happy
        case .sleeping: .sleeping
        case .grabbed: .grabbed
        }
    }

    private func scheduleDismiss() {
        dismissTask?.cancel()
        dismissTask = Task { @MainActor [weak self] in
            try? await Task.sleep(for: .seconds(6))
            guard Task.isCancelled == false else { return }
            self?.showsBubble = false
            self?.mood = .sleepy
            if self?.pose != .grabbed { self?.pose = .idle }
        }
    }

    private func beginActivity(_ newActivity: PetActivity, duration: Duration) {
        activityTask?.cancel()
        activity = newActivity
        animationSequenceIndex = 0
        walkingDistance = 0
        animationFrame = newActivity.frameSequence[0]
        activityTask = Task { @MainActor [weak self] in
            try? await Task.sleep(for: duration)
            guard Task.isCancelled == false, let self, self.pose != .grabbed else { return }
            self.activity = .idle
            self.pose = .idle
            self.mood = .sleepy
        }
    }

    private func interruptActivity() {
        activityTask?.cancel()
        activityTask = nil
        activity = .idle
        animationSequenceIndex = 0
        walkingDistance = 0
        animationFrame = 0
    }
}

File: windows/MutsumiPet.Windows/Stores/PetStore.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using MutsumiPet.Models;
using MutsumiPet.Support;

namespace MutsumiPet.Stores
{
    /// All of the pet's behaviour lives here, deliberately free of window and
    /// rendering types so the mood/activity state machine can be unit tested.
    public sealed class PetStore : INotifyPropertyChanged
    {
        public const string ScaleKey = "pet.scale";
        public const string LayerModeKey = "pet.windowLayerMode";
        public const string WanderingEnabledKey = "pet.wanderingEnabled";
        public const string WanderSpeedKey = "pet.wanderSpeed";

        public static readonly TimeSpan DismissDelay = TimeSpan.FromSeconds(6);
        public static readonly TimeSpan SettleDelay = TimeSpan.FromMilliseconds(120);
        public static readonly TimeSpan WalkTimeout = TimeSpan.FromSeconds(20);
        public static readonly TimeSpan SipDuration = TimeSpan.FromSeconds(7);

        private readonly IPetSettings settings;
        private readonly IPetScheduler scheduler;

        private PetMood mood = PetMood.Sleepy;
        private PetPose pose = PetPose.Idle;
        private string message;
        private bool showsBubble = true;
        private WindowLayerMode layerMode;
        private double scale;
        private PetActivity activity = PetActivity.Idle;
        private int walkingDirection = -1;
        private bool wanderingEnabled;
        private double wanderSpeed;
        private int animationFrame;

        private IDisposable dismissWork;
        private IDisposable settleWork;
        private IDisposable activityWork;
        private int animationSequenceIndex;
        private double walkingDistance;

        public PetStore(IPetSettings settings, IPetScheduler scheduler)
        {
            this.settings = settings;
            this.scheduler = scheduler;

            message = PetDialogue.Lines(PetDialogueScene.Idle)[0];

            double savedScale = settings.GetDouble(ScaleKey) ?? 1;
            scale = Math.Min(Math.Max(savedScale, 0.6), 1.4);
            layerMode = WindowLayerModes.Parse(settings.GetString(LayerModeKey), WindowLayerMode.Front);
            wanderingEnabled = settings.GetBool(WanderingEnabledKey) ?? true;
            double savedSpeed = settings.GetDouble(WanderSpeedKey) ?? 42;
            wanderSpeed = Math.Min(Math.Max(savedSpeed, 18), 90);
        }

        public event PropertyChangedEventHandler PropertyChanged;

        public PetMood Mood
        {
            get { return mood; }
            private set { SetField(ref mood, value, "Mood"); }
        }

        public PetPose Pose
        {
            get { return pose; }
            private set { SetField(ref pose, value, "Pose"); }
        }

        public string Message
        {
            get { return message; }
            private set { SetField(ref message, value, "Message"); }
        }

        public bool ShowsBubble
        {
            get { return showsBubble; }
            private set { SetField(ref showsBubble, value, "ShowsBubble"); }
        }

        public WindowLayerMode LayerMode
        {
            get { return layerMode; }
            private set { SetField(ref layerMode, value, "LayerMode"); }
        }

        public double Scale
        {
            get { return scale; }
            private set { SetField(ref scale, value, "Scale"); }
        }

        public PetActivity Activity
        {
            get { return activity; }
            private set { SetField(ref activity, value, "Activity"); }
        }

        public int WalkingDirection
        {
            get { return walkingDirection; }
            private set { SetField(ref walkingDirection, value, "WalkingDirection"); }
        }

        public bool WanderingEnabled
        {
            get { return wanderingEnabled; }
            private set { SetField(ref wanderingEnabled, value, "WanderingEnabled"); }
        }

        public double WanderSpeed
        {
            get { return wanderSpeed; }
            private set { SetField(ref wanderSpeed, value, "WanderSpeed"); }
        }

        public int AnimationFrame
        {
            get { return animationFrame; }
            private set { SetField(ref animationFrame, value, "AnimationFrame"); }
        }

        public void ReactToTap()
        {
            ReactToTap(null);
        }

        public void ReactToTap(int? randomDialogueIndex)
        {
            InterruptActivity();
            Mood = PetMoods.AllCases[((int)mood + 1) % PetMoods.AllCases.Length];
            switch (mood)
            {
                case PetMood.Curious: Pose = PetPose.Curious; break;
                case PetMood.Pleased: Pose = PetPose.Happy; break;
                default: Pose = PetPose.Idle; break;
            }
            Message = PetDialogue.Line(SceneFor(pose), randomDialogueIndex);
            ShowsBubble = true;
            ScheduleDismiss();
        }

        public void IdleTick()
        {
            IdleTick(null, null);
        }

        public void IdleTick(int? randomIndex, bool? sleeping)
        {
            if (showsBubble) return;
            bool usesSleepingPose = sleeping ?? PetRandom.Next(0, 2) == 1;
            PetDialogueScene scene = usesSleepingPose ? PetDialogueScene.Sleeping : PetDialogueScene.Idle;
            Message = PetDialogue.Line(scene, randomIndex);
            Mood = PetMood.Sleepy;
            Pose = usesSleepingPose ? PetPose.Sleeping : PetPose.Idle;
            ShowsBubble = true;
            ScheduleDismiss();
        }

        public void BeginDrag()
        {
            BeginDrag(null);
        }

        public void BeginDrag(int? randomDialogueIndex)
        {
            if (pose == PetPose.Grabbed) return;
            InterruptActivity();
            Cancel(ref settleWork);
            Cancel(ref dismissWork);
            Pose = PetPose.Grabbed;
            Mood = PetMood.Curious;
            Message = PetDialogue.Line(PetDialogueScene.Grabbed, randomDialogueIndex);
            ShowsBubble = true;
        }

        public void EndDrag()
        {
            EndDrag(null);
        }

        public void EndDrag(int? randomDialogueIndex)
        {
            if (pose != PetPose.Grabbed) return;
            Pose = PetPose.Curious;
            Message = PetDialogue.Line(PetDialogueScene.Released, randomDialogueIndex);
            ScheduleDismiss();
            Cancel(ref settleWork);
            settleWork = scheduler.Schedule(SettleDelay, delegate
            {
                if (pose == PetPose.Grabbed) return;
                Pose = PetPose.Idle;
                Mood = PetMood.Sleepy;
            });
        }

        public void ToggleBubble()
        {
            ShowsBubble = showsBubble == false;
            if (showsBubble) ScheduleDismiss();
        }

        public void SetScale(double newScale)
        {
            Scale = Math.Min(Math.Max(newScale, 0.6), 1.4);
            settings.SetDouble(ScaleKey, scale);
        }

        public void SetLayerMode(WindowLayerMode mode)
        {
            LayerMode = mode;
            settings.SetString(LayerModeKey, WindowLayerModes.RawValue(mode));
        }

        public void SetWanderingEnabled(bool enabled)
        {
            WanderingEnabled = enabled;
            settings.SetBool(WanderingEnabledKey, enabled);
            if (enabled == false) InterruptActivity();
        }

        public void SetWanderSpeed(double speed)
        {
            WanderSpeed = Math.Min(Math.Max(speed, 18), 90);
            settings.SetDouble(WanderSpeedKey, wanderSpeed);
        }

        public void LifestyleTick()
        {
            LifestyleTick(null, null);
        }

        public void LifestyleTick(int? randomEvent, int? randomDialogueIndex)
        {
            if (wanderingEnabled == false) return;
            if (pose == PetPose.Grabbed) return;
            if (activity != PetActivity.Idle) return;

            switch (randomEvent ?? PetRandom.Next(0, 8))
            {
                case 0:
                case 1:
                case 2:
                case 3:
                    PerformLifestyle(PetActivity.Walking, randomDialogueIndex);
                    break;
                case 4:
                    PerformLifestyle(PetActivity.DrinkingTea, randomDialogueIndex);
                    break;
                case 5:
                    PerformLifestyle(PetActivity.EatingSnack, randomDialogueIndex);
                    break;
                case 6:
                    Pose = PetPose.Sleeping;
                    Mood = PetMood.Sleepy;
                    Message = PetDialogue.Line(PetDialogueScene.Sleeping, randomDialogueIndex);
                    ShowsBubble = true;
                    ScheduleDismiss();
                    break;
                default:
                    Pose = PetPose.Curious;
                    Mood = PetMood.Curious;
                    Message = PetDialogue.Line(PetDialogueScene.Curious, randomDialogueIndex);
                    ShowsBubble = true;
                    ScheduleDismiss();
                    break;
            }
        }

        public void PerformLifestyle(PetActivity requestedActivity)
        {
            PerformLifestyle(requestedActivity, null);
        }

        public void PerformLifestyle(PetActivity requestedActivity, int? randomDialogueIndex)
        {
            if (wanderingEnabled == false) return;
            if (pose == PetPose.Grabbed) return;
            InterruptActivity();

            switch (requestedActivity)
            {
                case PetActivity.Walking:
                    // Arrival normally ends walking; this timeout is only a safety net
                    // for a window that cannot move or disappears mid-activity.
                    BeginActivity(PetActivity.Walking, WalkTimeout);
                    Pose = PetPose.Curious;
                    Mood = PetMood.Curious;
                    Message = PetDialogue.Line(PetDialogueScene.Walking, randomDialogueIndex);
                    ShowsBubble = true;
                    ScheduleDismiss();
                    break;
                case PetActivity.DrinkingTea:
                    BeginActivity(PetActivity.DrinkingTea, SipDuration);
                    Pose = PetPose.Idle;
                    Mood = PetMood.Pleased;
                    Message = PetDialogue.Line(PetDialogueScene.DrinkingTea, randomDialogueIndex);
                    ShowsBubble = true;
                    ScheduleDismiss();
                    break;
                case PetActivity.EatingSnack:
                    BeginActivity(PetActivity.EatingSnack, SipDuration);
                    Pose = PetPose.Happy;
                    Mood = PetMood.Pleased;
                    Message = PetDialogue.Line(PetDialogueScene.EatingSnack, randomDialogueIndex);
                    ShowsBubble = true;
                    ScheduleDismiss();
                    break;
            }
        }

        public void UpdateWalkingDirection(int direction)
        {
            WalkingDirection = direction < 0 ? -1 : 1;
        }

        public void AdvanceActivityFrame()
        {
            if (activity == PetActivity.Idle) return;
            int[] sequence = PetActivities.FrameSequence(activity);
            animationSequenceIndex = (animationSequenceIndex + 1) % sequence.Length;
            AnimationFrame = sequence[animationSequenceIndex];
        }

        public void AdvanceWalkingFrame(double distance)
        {
            if (activity != PetActivity.Walking) return;
            if (distance <= 0) return;
            walkingDistance += distance;
            double strideLength = 44 * scale;
            double phase = walkingDistance % strideLength / strideLength;
            AnimationFrame = Math.Min((int)(phase * 4), 3);
        }

        public void FinishWalking()
        {
            if (activity != PetActivity.Walking) return;
            InterruptActivity();
            Pose = PetPose.Idle;
            Mood = PetMood.Sleepy;
            Message = PetDialogue.Line(PetDialogueScene.Idle);
            ShowsBubble = true;
            ScheduleDismiss();
        }

        public void SpeakForCurrentState()
        {
            SpeakForCurrentState(null);
        }

        public void SpeakForCurrentState(int? randomDialogueIndex)
        {
            Message = PetDialogue.Line(CurrentDialogueScene, randomDialogueIndex);
            ShowsBubble = true;
            ScheduleDismiss();
        }

        private PetDialogueScene CurrentDialogueScene
        {
            get
            {
                switch (activity)
                {
                    case PetActivity.Walking: return PetDialogueScene.Walking;
                    case PetActivity.DrinkingTea: return PetDialogueScene.DrinkingTea;
                    case PetActivity.EatingSnack: return PetDialogueScene.EatingSnack;
                    default: return SceneFor(pose);
                }
            }
        }

        private static PetDialogueScene SceneFor(PetPose pose)
        {
            switch (pose)
            {
                case PetPose.Curious: return PetDialogueScene.Curious;
                case PetPose.Happy: return PetDialogueScene.Happy;
                case PetPose.Sleeping: return PetDialogueScene.Sleeping;
                case PetPose.Grabbed: return PetDialogueScene.Grabbed;
                default: return PetDialogueScene.Idle;
            }
        }

        private void ScheduleDismiss()
        {
            Cancel(ref dismissWork);
            dismissWork = scheduler.Schedule(DismissDelay, delegate
            {
                ShowsBubble = false;
                Mood = PetMood.Sleepy;
                if (pose != PetPose.Grabbed) Pose = PetPose.Idle;
            });
        }

        private void BeginActivity(PetActivity newActivity, TimeSpan duration)
        {
            Cancel(ref activityWork);
            Activity = newActivity;
            animationSequenceIndex = 0;
            walkingDistance = 0;
            AnimationFrame = PetActivities.FrameSequence(newActivity)[0];
            activityWork = scheduler.Schedule(duration, delegate
            {
                if (pose == PetPose.Grabbed) return;
                Activity = PetActivity.Idle;
                Pose = PetPose.Idle;
                Mood = PetMood.Sleepy;
            });
        }

        private void InterruptActivity()
        {
            Cancel(ref activityWork);
            Activity = PetActivity.Idle;
            animationSequenceIndex = 0;
            walkingDistance = 0;
            AnimationFrame = 0;
        }

        private static void Cancel(ref IDisposable work)
        {
            if (work == null) return;
            work.Dispose();
            work = null;
        }

        private void SetField<T>(ref T field, T value, string propertyName)
        {
            if (EqualityComparer<T>.Default.Equals(field, value)) return;
            field = value;
            PropertyChangedEventHandler handler = PropertyChanged;
            if (handler != null) handler(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}
